api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+  // defined in [ExecutorInput](). The output of the container follows the
+  // contract of [ExecutorOutput]().
+  message PipelineContainerSpec {
+    // The image uri of the container.
+    string image = 1;
+    // The main entrypoint commands of the container to run. If not provided,
+    // fallback to use the entry point command defined in the container image.
+    repeated string command = 2;
+    // The arguments to pass into the main entrypoint of the container.
+    repeated string args = 3;
+
+    // The lifecycle hooks of the container.
+    // Each hook follows the same I/O contract as the main container entrypoint.
+    // See [ExecutorInput]() and [ExecutorOutput]() for details.
+    // (-- TODO(b/165323565): add more documentation on caching and lifecycle
+    // hooks. --)
+    message Lifecycle {
+      // The command and args to execute a program.
+      message Exec {
+        // The command of the exec program.
+        repeated string command = 2;
+        // The args of the exec program.
+        repeated string args = 3;
+      }
+      // This hook is invoked before caching check. It can change the properties
+      // of the execution and output artifacts before they are used to compute
+      // the cache key. The updated metadata will be passed into the main
+      // container entrypoint.
+      Exec pre_cache_check = 1;
+    }
+    // The lifecycle hooks of the container executor.
+    Lifecycle lifecycle = 4;
+  }
+
+  // The specification to import or reimport a new artifact to the pipeline.
+  message ImporterSpec {
+    // The URI of the artifact.
+    string artifact_uri = 1;
+
+    // The type of the artifact.
+    ArtifactTypeSchema type_schema = 2;
+
+    // The properties of the artifact.
+    map<string, ValueOrRuntimeParameter> properties = 3;
+
+    // The custom properties of the artifact.
+    map<string, ValueOrRuntimeParameter> custom_properties = 4;
+
+    // Whether or not import an artifact regardless it has been imported before.
+    bool reimport = 5;
+  }
+
+  // ResolverSpec is subject to change. Currently we only use enum to
+  // represent two of the currently available policies. We plan to introduce a
+  // flexible config to enable more sophisticated policies in the future.
+  // TODO(b/152230663): Support more flexibility for resolution logic.
+  message ResolverSpec {
+    enum ResolverPolicy {
+      RESOLVER_POLICY_UNSPECIFIED = 0;
+      // Within the models associated with the same pipeline context, always
+      // select the model with largest `last_update_time_since_epoch`.
+      LATEST_MODEL = 1;
+      // Within the models associated with the same pipeline context,
+      // select the latest Model that is blessed (represented by having a
+      // corresponding ModelBlessing artifact. The result will be consumed by
+      // components that need the latest model that is blessed such as
+      // Evaluator.
+      LATEST_BLESSED_MODEL = 2;

```

Comment:
The ResolverPolicy seems to be pretty specialized.
Can we keep the schema more generic?

Reply:
Yes, we have a TODO for it. MLMD team is working on a language to build more flexible queries. /cc @neuromage